model/modelprocessor/hostname.go

```diff
@@ -29,25 +29,25 @@ import (
 type SetHostHostname struct{}
 
 // ProcessBatch sets or overrides the host.name and host.hostname fields for events.
-func (SetHostHostname) ProcessBatch(ctx context.Context, b *model.Batch) error {
+func (SetHostHostname) ProcessBatch(ctx context.Context, b *modelpb.Batch) error {
 	for i := range *b {
-		setHostHostname(&(*b)[i])
+		setHostHostname((*b)[i])
 	}
 	return nil
 }
 
-func setHostHostname(event *model.APMEvent) {
+func setHostHostname(event *modelpb.APMEvent) {
 	switch {
-	case event.Kubernetes.NodeName != "":
+	case event.GetKubernetes().GetNodeName() != "":
 		// host.kubernetes.node.name is set: set host.hostname to its value.
 		event.Host.Hostname = event.Kubernetes.NodeName
-	case event.Kubernetes.PodName != "" || event.Kubernetes.PodUID != "" || event.Kubernetes.Namespace != "":
+	case event.GetKubernetes().GetPodName() != "" || event.GetKubernetes().GetPodUid() != "" || event.GetKubernetes().GetNamespace() != "":
 		// kubernetes.* is set, but kubernetes.node.name is not: don't set host.hostname at all.
 		event.Host.Hostname = ""
 	default:
 		// Otherwise use the originally specified host.hostname value.
 	}
-	if event.Host.Name == "" {
+	if event.GetHost().GetName() == "" && event.GetHost().GetHostname() != "" {
 		event.Host.Name = event.Host.Hostname
```

Comment:
We're setting the hostname here and above without checking if `event.Host != nil`, something like this is safer:

```go
func setHostHostname(event *modelpb.APMEvent) {
	k8s := event.GetKubernetes()
	if event.GetHost() == nil {
		event.Host = &modelpb.Host{}
	}
	switch {
	case k8s.GetNodeName() != "":
		// host.kubernetes.node.name is set: set host.hostname to its value.
		event.Host.Hostname = k8s.NodeName
	case k8s.GetPodName() != "" || k8s.GetPodUid() != "" || k8s.GetNamespace() != "":
		// kubernetes.* is set, but kubernetes.node.name is not: don't set host.hostname at all.
		event.Host.Hostname = ""
	default:
		// Otherwise use the originally specified host.hostname value.
	}
	if host := event.GetHost(); host.GetName() == "" && host.GetHostname() != "" {
		event.Host.Name = event.Host.Hostname
	}
}
```

Reply:
In this case `event.GetHost().GetHostname() != ""` guarantees that `event.Host` is not nil 

For the switch case above: I agree, good catch. We still need to make sure to only set the host if the hostname will actually be set

Reply:
Forgot to leave a comment: I've updated the code

Reply:
It doesn't look right, the `event.Host` is being set inside the switch case, which may or may not be set. In cases where the Kubernetes and event.Host structs are nil, line 56 will panic.

Reply:
I think I'm missing something. Line 56 is guarded by:

`if event.GetHost().GetName() == "" && event.GetHost().GetHostname() != "" {`

The only way for `event.GetHost().GetHostname() != ""` to be true is if `event.Host` is not nil so it shouldn't be possible for line 56 to panic.

Reply:
Nothing is missing. I misread the second part of the if clause.